subsys/event_manager/event_manager.c

```diff
@@ -0,0 +1,237 @@
+{
+	if (CONFIG_EVENT_MANAGER_EVENT_LOG_BUF_LEN < 2) {
+		return;
+	}
+
+	BUILD_ASSERT(CONFIG_EVENT_MANAGER_EVENT_LOG_BUF_LEN != 1);
+	char log_buf[CONFIG_EVENT_MANAGER_EVENT_LOG_BUF_LEN];
+
+	int pos = et->log_event_func_dep(eh, log_buf, sizeof(log_buf));
+
+	if (pos < 0) {
+		log_buf[0] = '\0';
+	} else if (pos >= sizeof(log_buf)) {
+		log_buf[sizeof(log_buf) - 2] = '~';
+	}
+
+	if (IS_ENABLED(CONFIG_EVENT_MANAGER_LOG_EVENT_TYPE)) {
+		LOG_INF("e: %s %s", et->name, log_strdup(log_buf));
+	} else {
+		LOG_INF("%s", log_strdup(log_buf));
+	}
+}
+
+static void log_event(const struct event_header *eh)
+{
+	const struct event_type *et = eh->type_id;
+
+	if (!IS_ENABLED(CONFIG_EVENT_MANAGER_SHOW_EVENTS) ||
+	    !log_is_event_displayed(et)) {
+		return;
+	}
+
+	if (et->log_event_func_dep) {
+		log_event_using_buffer(eh, et);
+	} else if (et->log_event_func) {
+		et->log_event_func(eh);
+	} else if (IS_ENABLED(CONFIG_EVENT_MANAGER_LOG_EVENT_TYPE)) {
+		LOG_INF("e: %s", et->name);
+	}
+}
+
+static void log_event_progress(const struct event_type *et,
+			       const struct event_listener *el)
+{
+	if (!IS_ENABLED(CONFIG_EVENT_MANAGER_SHOW_EVENTS) ||
+	    !IS_ENABLED(CONFIG_EVENT_MANAGER_SHOW_EVENT_HANDLERS) ||
+	    !log_is_event_displayed(et)) {
+		return;
+	}
+
+	LOG_INF("|\tnotifying %s", el->name);
+}
+
+static void log_event_consumed(const struct event_type *et)
+{
+	if (!IS_ENABLED(CONFIG_EVENT_MANAGER_SHOW_EVENTS) ||
+	    !IS_ENABLED(CONFIG_EVENT_MANAGER_SHOW_EVENT_HANDLERS) ||
+	    !log_is_event_displayed(et)) {
+		return;
+	}
+
+	LOG_INF("|\tevent consumed");
+}
+
+static void log_event_init(void)
+{
+	if (!IS_ENABLED(CONFIG_LOG)) {
+		return;
+	}
+
+	STRUCT_SECTION_FOREACH(event_type, et) {
+		if (et->init_log_enable) {
+			size_t idx = et - _event_type_list_start;
+
+			atomic_set_bit(_event_manager_event_display_bm.flags, idx);
+		}
+	}
+}
+
+void __weak event_manager_trace_event_execution(const struct event_header *eh,
+				  bool is_start)
+{
+}
+
+void __weak event_manager_trace_event_submission(const struct event_header *eh,
+				const void *trace_info)
+{
+}
+
+void * __weak event_manager_alloc(size_t size)
+{
+	void *event = k_malloc(size);
+
+	if (unlikely(!event)) {
+		LOG_ERR("Event Manager OOM error\n");
+		__ASSERT_NO_MSG(false);
+		#if defined(CONFIG_REBOOT)
+			sys_reboot(SYS_REBOOT_WARM);
+		#else
+			k_panic();
+		#endif
+		return NULL;
+	}
+
+	return event;
+}
+
+void __weak event_manager_free(void *addr)
+{
+	k_free(addr);
+}
+
+int __weak event_manager_trace_event_init(void)
+{
+	return 0;
+}
+
+static void event_processor_fn(struct k_work *work)
+{
+	sys_slist_t events = SYS_SLIST_STATIC_INIT(&events);
+
+	/* Make current event list local. */
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	if (sys_slist_is_empty(&eventq)) {
+		k_spin_unlock(&lock, key);
+		return;
+	}
+
+	sys_slist_merge_slist(&events, &eventq);
+
+	k_spin_unlock(&lock, key);
```

Comment:
The same can be implemented in one-line using `struct k_fifo` and `k_fifo_get()`. I also see no advantages not to do it this way, very unlikely `eventq` would contain more than a single event per `k_work_submit(&event_processor);`.

Reply:
Maybe. It could be a potential change for later.
Originally it was implemented to be as fast as possible so heavier api was avoided. To change it we would need to see if there would be no impact on performance.